### PR TITLE
LaunchOptions, Fetch Domain, authenticate()

### PR DESCRIPTION
**Pull request by jeprojects**
Added the ability to include command line args.
Example: LaunchOptionsBuilder::default().args(vec!["--proxy-server=127.0.0.1:8888"])

Added Fetch domain, with methods: enable_fetch(), disable_fetch(), enable_request_interception(), authenticate()

Added network::SetExtraHttpHeaders protocol and set_extra_http_headers()
Added network::SetCookies protocol and set_cookies() method

### CODE REVIEW

**Comment by atroche** [comment]
Wow, this is exciting! Sorry it took me so long to come around to this. Any outstanding problems or TODOs you're aware of, before I dive into the code? Any design decisions you're unsure about? How about breaking changes to the public API?

**Comment by jeprojects** [comment]
No TODOs. All though I have tried to base the design decisions on what you are currently doing, please take a look at the RequestInterceptor trait. 

Open to suggestions for improvement if needed.

No breaking changes.

**Comment by Dowwie** [comment]
@atroche Please review and approve this PR.  This is a huge improvement in the API with much needed functionality.

@jeprojects Thanks for this PR!  I tried working on setExtraHTTPHeaders and struggled for far too long, discovering from your commit that I was missing  the initial call to "Enable" :`self.call_method(network::methods::Enable {})?;`

File: src/browser/process.rs [hunk 1]
```diff
@@ -233,6 +238,15 @@ impl Process {
 
         args.extend(&DEFAULT_ARGS);
 
+        if !launch_options.args.is_empty() {
+            let extra_args: Vec<&str> = launch_options
+                .args
+                .iter()
+                .map(|a| a.to_str().unwrap())
```
Comment (leshow) [inline src/browser/process.rs]: Could `flat_map` this to get rid of the `unwrap()`

Comment (leshow) [reply src/browser/process.rs]: You probably don't even need to `collect`. You can just pass the iterator to `extend`

File: src/browser/tab/mod.rs [hunk 2]
```diff
@@ -10,42 +10,36 @@ use serde;
...
 mod keys;
 mod point;
 
 #[derive(Debug)]
-pub enum RequestInterceptionDecision {
-    Continue,
-    // TODO: Error
-    Response(String),
+pub enum RequestPausedDecision {
+    Fulfil(fetch::methods::FulfilRequest),
```
Comment (leshow) [inline src/browser/tab/mod.rs]: Fulfill*

File: src/browser/tab/mod.rs [hunk 3]
```diff
@@ -58,6 +52,31 @@ pub type ResponseHandler = Box<
...
+    fn intercept(
+        &self,
+        transport: Arc<Transport>,
+        session_id: SessionId,
+        event: RequestPausedEvent,
+    ) -> RequestPausedDecision;
+}
+
+impl<
+        F: Fn(Arc<Transport>, SessionId, RequestPausedEvent) -> RequestPausedDecision + Send + Sync,
```
Comment (leshow) [inline src/browser/tab/mod.rs]: super small nit but you may want to add this in a `where` clause for readability

File: src/browser/tab/mod.rs [hunk 4]
```diff
@@ -58,6 +52,31 @@ pub type ResponseHandler = Box<
...
+    > RequestInterceptor for F
+{
+    fn intercept(
+        &self,
+        transport: Arc<Transport>,
+        session_id: SessionId,
+        event: RequestPausedEvent,
+    ) -> RequestPausedDecision {
+        self(transport, session_id, event)
```
Comment (leshow) [inline src/browser/tab/mod.rs]: I'm surprised this works without explicitly `(self)(...)`. Cool

File: src/browser/tab/mod.rs [hunk 5]
```diff
@@ -227,38 +252,53 @@ impl<'a> Tab {
...
+                            RequestPausedDecision::Fail(fail_request) => transport
+                                .call_method_on_target(session_id.clone(), fail_request)
+                                .map(|_| ()),
+                        };
+                        if result.is_err() {
+                            warn!("Tried to handle request after connection was closed");
+                        }
+                    }
+                    Event::AuthRequired(event) => {
+                        let auth_challenge_response = auth_handler_mutex.lock().unwrap().clone();
```
Comment (leshow) [inline src/browser/tab/mod.rs]: `unwrap` here? If you want to bail on the `lock` failing, you can use `expect`

File: src/browser/tab/mod.rs [hunk 6]
```diff
@@ -723,47 +763,50 @@ impl<'a> Tab {
...
+        self.call_method(fetch::methods::Enable {
+            patterns,
+            handle_auth_requests,
+        })?;
+        Ok(self)
+    }
+
+    /// Disables fetch domain
+    pub fn disable_fetch(&self) -> Fallible<&Self> {
+        self.call_method(fetch::methods::Disable {})?;
```
Comment (leshow) [inline src/browser/tab/mod.rs]: If `Disable` has no members you can leave off it's `{}` in it's definition and creation here.

File: src/browser/tab/mod.rs [hunk 7]
```diff
@@ -723,47 +763,50 @@ impl<'a> Tab {
...
-    /// The closure needs to return a variant of `RequestInterceptionDecision` (so, `Continue` or
-    /// `Response(String)`).
-    pub fn enable_request_interception(
-        &self,
-        patterns: &[network::methods::RequestPattern],
-        interceptor: RequestInterceptor,
-    ) -> Fallible<()> {
+    /// The closure needs to return a variant of `RequestPausedDecision`.
+    pub fn enable_request_interception(&self, interceptor: Arc<RequestIntercept>) -> Fallible<()> {
         let mut current_interceptor = self.request_interceptor.lock().unwrap();
```
Comment (leshow) [inline src/browser/tab/mod.rs]: Bubble up the error in your `unwrap()`

File: src/browser/tab/mod.rs (hunk 6 above)
Comment (leshow) [reply src/browser/tab/mod.rs]: (I see this is how the rest of the code is written, so feel free to ignore)

File: tests/simple.rs [hunk 8]
```diff
@@ -489,58 +494,66 @@ fn set_request_interception() -> Fallible<()> {
...
-                let base64_response = base64::encode(&wrapped_writer);
-
-                RequestInterceptionDecision::Response(base64_response)
+
+                let headers = vec![HeaderEntry {
+                    name: "Content-Type".to_string(),
+                    value: "application/javascript".to_string(),
+                }];
+
+                let fulful_request = FulfilRequest {
```
Comment (leshow) [inline tests/simple.rs]: This is a fulfilled request? May want to change the name of this var

**Comment by Dowwie** [comment]
thanks @leshow  !  @jeprojects note his review

**Comment by Dowwie** [comment]
@jeprojects ping.

**Comment by jeprojects** [comment]
@leshow Thanks for the feedback. I will make the changes necessary.

**Review by mdrokz — APPROVED** [review]

**Pull request merged**